ocaml/xapi/cert_distrib.ml

```diff
@@ -364,15 +364,32 @@ let collect_pool_certs ~__context ~rpc ~session_id ~map ~from_hosts =
          map cert
      )
 
-let take_and_append n x xs =
-  (* take_and_append 3 10 [1;2;3;4] = [1;2;3;10] *)
-  let rec loop i acc = function
-    | x :: xs when i < n ->
-        loop (i + 1) (x :: acc) xs
-    | _ ->
-        x :: acc |> List.rev
+(* insert an element randomly *between* existing elements of a list.
+   We are enumerating the gaps between elements in a list:
+
+   List:     a b c d
+   Gaps:    0 1 2 3 4
+
+   A list of length n has gaps 0, 1, ...,n. We are selecting a gap
+   between 1 and n-1 to insert an element - so never at the beginning or
```

Comment:
Why not at the beginning for n >= 2, while it is allowed for n = 1?

Reply:
Unless you drop the element to be inserted, you don't have a choice when you have `[]` or `[a]`. I believe this edge case should never come up but I want to handle it reasonably.

Reply:
The old function allows adding the element to the beginning for any length. I think it's valid in the context where we use it, making the call fail right away.